### ISSUE TITLE
Inverter Serial Number

### ISSUE THREAD

**Issue by andrerdrummond** [tbnobody/OpenDTU#199]
To make openDTU starts to work with my ~HM-1500~ MI-1500, I needed to change HM_4CH.cpp 
**From**: return serial >= 0x116100000000 && serial <= 0x116199999999;
**To**: 0x106100000000 && serial <= 0x116199999999;
Because the serial number starts with 1062XXXXXXX. I suggest to put an information at readme for other people.

**Comment by tbnobody** [comment]
it's not as easy.... for all inverters which I have seen at the moment, serials beginning with 1061 are MI- types... 1161 is used to identify HM- types... if you really have an HM- type with this serial number you are the first one.

**Comment by stefan123t** [comment]
Whoa, this looks like an overflow of the year index on the fifth :smile: 
With the 5th digit starting with `1` in 2015 we expected it to reach `9` in 2024.
Currently built HM- inverters have 1161**8**CWxxxxx for 2022 and last years had 1161**7**CWxxxxx for 2021.
While CW stands for the calendar week the inverter was manufactured.

@andrerdrummond is this a special version e.g. a HM-1500T (trough connector model) or built for a different market.
The developers are mostly using european / german / austrian market models.

**Comment by stefan123t** [comment]
@andrerdrummond can you provide us a photo of the inverter with the serial number sticker ?
You may blank the last five digits if you really mind, however we would like to witness the first 4+1+2 digits (model+year+calendar week) and the overall look of your HM-1500.

**Comment by andrerdrummond** [comment]
> it's not as easy.... for all inverters which I have seen at the moment, serials beginning with 1061 are MI- types... 1161 is used to identify HM- types... if you really have an HM- type with this serial number you are the first one. (Protocoll of MI is different compared to HM)

Sorry, mine is a MI-1500 like you said. But with the change that I made like a said, openDTU is reading something.

**Comment by andrerdrummond** [comment]
> @andrerdrummond can you provide us a photo of the inverter with the serial number sticker ? You may blank the last five digits if you really mind, however we would like to witness the first 4+1+2 digits (model+year+calendar week) and the overall look of your HM-1500.

![Mi-1500](https://user-images.githubusercontent.com/11406518/194647887-062feddd-80fb-462c-b115-a6c1bb9091d8.JPG)

**Comment by stefan123t** [comment]
@andrerdrummond your inverter is a 1062 PreId 7 (buildyear =2021) 34 (buildweek=34)
The PreId can be piped through the UsartNrf_GetInvterType() Method of the original DTU Pro source code.

@tbnobody it should fall into this condition:
```c
        else if(((pId[1] & 0xf0) ==  0x50) || ((pId[1] & 0xf0) ==  0x60)) // 0x0050 || 0x0060
        {
            if(((pId[0] == 0x10) && (pId[1] == 0x62)) || ((pId[0] == 0x11) && (pId[1] == 0x61))) // 0x1062 || 0x1161
            {
                return Inverter_HM_OneToFour; // 10*300W=3000W
            }
            else if((pId[0] == 0x10) && (pId[1] == 0x69)) // 0x1069
            {
                return InitInverterType; // 0*300W=0W
            }
            else // others
            {
                return Inverter_1000; // 4*300W=1200W
            }
        }
```

**Comment by stefan123t** [comment]
@tbnobody did you notice that in the example above ^^^ the options which match the 0x0050 || 0x0060 precondition are actually
1. only the 0x1062 || 0x1161 are HM_OneToFour model. 
   Here the 0x1062 is MI-1500 which talks the same Gen3 Protocol as the proper HM-1500.
2. 0x1069 that is InitInverterType
   I am unsure about this, something like a reboot or DTU does not know itself ?
3. and anything with a 0x0050 is (MI-)1000
   This is most likely a Gen2 Protocol type if I recall correctly.

That is I do not fully understand the complex changes you made in 223b7ff and how you arrived at your conclusion for the following code: 

```c
    uint8_t preId[2];
    preId[0] = (uint8_t)(serial >> 40);
    preId[1] = (uint8_t)(serial >> 32);

    if ((uint8_t)(((((uint16_t)preId[0] << 8) | preId[1]) >> 4) & 0xff) == 0x16) {  // why is this necessary ?
        return true;
    }

    if ((((preId[1] & 0xf0) == 0x50) || ((preId[1] & 0xf0) == 0x60))  // this could be simplified for us
        && (((preId[0] == 0x10) && (preId[1] == 0x62)) || ((preId[0] == 0x11) && (preId[1] == 0x61)))) {
        return true;
    }
```

IMHO it should be sufficient for our purpose to check for 0x1062 || 0x1161 being HM_OneToFour model only.

Why are you checking this strange pre-condition too (which is only valid for 3. which we are not interested in IMHO)
and why are we returning true within two conditions ?
Do we really expect something like a 0x2**16**9 to be a HM_OneToFour model too which would match our first condition ?

**Comment by andrerdrummond** [comment]
> It should be fixed with [223b7ff](https://github.com/tbnobody/OpenDTU/commit/223b7ff3ce0da7dd7d95931551d42d643cb99937) @andrerdrummond could you please doublecheck?

Updated the OpenDTU and uploaded, it's working.
